src/stream.rs

```diff
@@ -43,117 +43,165 @@
-    /// Check the socket's read readiness state.
-    pub fn poll_read_ready(&self, cx: &mut Context, mask: Ready) -> Poll<Result<Ready>> {
-        self.io.poll_read_ready(cx, mask)
-    }
+        loop {
+            let stream = unsafe { vsock::VsockStream::from_raw_fd(socket) };
+            let stream = Self::new(stream)?;
+            let mut guard = stream.inner.writable().await?;
 
-    /// Check the socket's write readiness state.
-    pub fn poll_write_ready(&self, cx: &mut Context) -> Poll<Result<Ready>> {
-        self.io.poll_write_ready(cx)
+            match guard.try_io(|_| Ok(())) {
+                Ok(_) => return Ok(stream),
+                Err(_would_block) => continue,
+            }
+        }
     }
 
     /// The local address that this socket is bound to.
     pub fn local_addr(&self) -> Result<SockAddr> {
-        self.io.get_ref().local_addr()
+        self.inner.get_ref().local_addr()
     }
 
     /// The remote address that this socket is connected to.
     pub fn peer_addr(&self) -> Result<SockAddr> {
-        self.io.get_ref().peer_addr()
+        self.inner.get_ref().peer_addr()
     }
 
     /// Shuts down the read, write, or both halves of this connection.
     pub fn shutdown(&self, how: Shutdown) -> Result<()> {
-        self.io.get_ref().shutdown(how)
+        self.inner.get_ref().shutdown(how)
     }
 
     pub(crate) fn poll_write_priv(&self, cx: &mut Context<'_>, buf: &[u8]) -> Poll<Result<usize>> {
-        ready!(self.io.poll_write_ready(cx))?;
-
-        match self.io.get_ref().write(buf) {
-            Err(ref e) if e.kind() == ErrorKind::WouldBlock => {
-                self.io.clear_write_ready(cx)?;
-                Poll::Pending
+        loop {
+            let mut guard = ready!(self.inner.poll_write_ready(cx))?;
+
+            match guard.try_io(|inner| inner.get_ref().write(buf)) {
+                Ok(result) => match result {
+                    Ok(n) => return Ok(n).into(),
+                    Err(e) => return Err(e).into(),
+                },
+                Err(_would_block) => continue,
             }
-            x => Poll::Ready(x),
         }
     }
 
     pub(crate) fn poll_read_priv(
         &self,
         cx: &mut Context<'_>,
-        buf: &mut [u8],
-    ) -> Poll<Result<usize>> {
-        ready!(self.io.poll_read_ready(cx, Ready::readable()))?;
-
-        match self.io.get_ref().read(buf) {
-            Err(ref e) if e.kind() == ErrorKind::WouldBlock => {
-                self.io.clear_read_ready(cx, Ready::readable())?;
-                Poll::Pending
+        buf: &mut ReadBuf<'_>,
+    ) -> Poll<Result<()>> {
+        let b;
+        unsafe {
+            b = &mut *(buf.unfilled_mut() as *mut [mem::MaybeUninit<u8>] as *mut [u8]);
+        };
+
+        loop {
+            let mut guard = ready!(self.inner.poll_read_ready(cx))?;
+
+            match guard.try_io(|inner| inner.get_ref().read(b)) {
+                Ok(result) => match result {
+                    Ok(n) => {
```

Comment:
double match is a bit hard to read.
How do you feel about `if let Some` instead?
```rust
            if let Some(result) = guard.try_io(|inner| inner.get_ref().read(b)) {
                return match result {
                    Ok(n) => {
                        unsafe {
                            buf.assume_init(n);
                        }
                        buf.advance(n);
                        Ok(()).into()
                    }
                    Err(e) => Err(e).into(),
                };
            }
```

Reply:
@mxpv Thank you very much for your review and good advice.

I have changed the code to this 
```
match guard.try_io(|inner| inner.get_ref().read(b)) {
    Ok(Ok(n)) => {
        unsafe {
            buf.assume_init(n);
        }
        buf.advance(n);
        return Ok(()).into();
    }
    Ok(Err(e)) => return Err(e).into(),
    Err(_would_block) => {
        continue;
    }
}
```

What about this style?

Reply:
nice, didn't know that pattern matcher in rust that smart :)